### PR TITLE
nixos/networkd: allow specifying FirewallMark mask

### PR DESCRIPTION
**Pull request by frantisekhanzlikbl**
## Description of changes

systemd-networkd permits the `routingPolicyRules` `FirewallMark` option to either be a single integer specifying an exact match, or a slash-separated pair of a value and a mask. Previously, the networkd module only accepted the first case. This PR adds support for the second case.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Ma27 — APPROVED** [review]
LGTM.
Will give the systemd maintainer team a few days to review, but I'd merge in about a week.

File: nixos/lib/systemd-lib.nix [hunk 1]
```diff
@@ -156,6 +160,23 @@ in rec {
     optional (attr ? ${name} && !(((isInt attr.${name} || isFloat attr.${name}) && min <= attr.${name} && max >= attr.${name}) || elem attr.${name} values))
       "Systemd ${group} field `${name}' is not a value in range [${toString min},${toString max}], or one of ${toString values}";
 
+  assertRangeWithOptionalMask = name: min: max: group: attr:
+    if (attr ? ${name}) then
+      if isInt attr.${name} then
+        assertRange name min max group attr
+      else if isString attr.${name} then
+        let
+          fields = match "([0-9]+)(/([0-9]+))?" attr.${name};
```
Comment (andrevmatos) [inline nixos/lib/systemd-lib.nix:169]: Could this be made to also support the hex notation? These can be really handy when specifying bitmasks, and is supported by systemd

**Comment by frantisekhanzlikbl** [comment]
Sure, I'll update the PR tomorrow. I actually thought about it, but since the the manpage doesn't explicitly mention that it's supported and I didn't need it at the time, I assumed it is not.
Thanks for the review. :-)
-------- Original Message --------On 24/07/2024 23:48, André Vitor de Lima Matos  wrote:
@andrevmatos commented on this pull request.



In nixos/lib/systemd-lib.nix:
> @@ -156,6 +160,23 @@ in rec {
     optional (attr ? ${name} && !(((isInt attr.${name} || isFloat attr.${name}) && min <= attr.${name} && max >= attr.${name}) || elem attr.${name} values))
       "Systemd ${group} field `${name}' is not a value in range [${toString min},${toString max}], or one of ${toString values}";

+  assertRangeWithOptionalMask = name: min: max: group: attr:
+    if (attr ? ${name}) then
+      if isInt attr.${name} then
+        assertRange name min max group attr
+      else if isString attr.${name} then
+        let
+          fields = match "([0-9]+)(/([0-9]+))?" attr.${name};

Could this be made to also support the hex notation? These can be really handy when specifying bitmasks, and is supported by systemd

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/NixOS/nixpkgs/pull/329262#pullrequestreview-2197851858",
"url": "https://github.com/NixOS/nixpkgs/pull/329262#pullrequestreview-2197851858",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by frantisekhanzlikbl** [comment]
sorry, it took me a bit longer to get back to this than I hoped it would.

regarding the hex support: is there a function in nixpkgs that takes an "integer string" and parses it in the base specified by its prefix? I could mix `toIntBase10` with `trivial.fromHexString` somewhere in the file or perform the same dirty trick that [`fromHexString`](https://github.com/NixOS/nixpkgs/blob/95688d40d6fb21b5e8509b5517710b93e0a769d0/lib/trivial.nix#L1098) does to parse the number, but this feels like a fairly common thing that should already have a ready-made function, alas I was unable to find none.

**Comment by andrevmatos** [comment]
I don't know about any, but that'd be needed only if you want to enforce the range or some logic with the mask. I'm not sure we need to go this deep in our validations, and just ensuring int10 or hex numbers (with regex) in both sides would already be enough IMO.

**Comment by frantisekhanzlikbl** [comment]
Hmm, I'm torn between "this is annoying & unlikely to happen, so let's let networkd handle it" and "this is networking related, and if it makes networkd not set up some important interface somewhere, no one's going to be happy". I guess I'll give it a try and see whether the implementation would be *that* annoying.

**Comment by frantisekhanzlikbl** [comment]
Alright, it was absolutely trivial, so unless someone feels strongly about the use of `fromTOML`, I think this is fine. :slightly_smiling_face:

**Review by fpletz — APPROVED** [review]

**Review by Ma27 — APPROVED** [review]

**Pull request merged**